### PR TITLE
Collapsible streams

### PR DESCRIPTION
**Pull request by dizzyjaguar**

### CODE REVIEW

**Comment by chrisbobbe** [comment]
Thanks, @dizzyjaguar!

- Could you please post some [screenshots](https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html), and perhaps a GIF?
- I see CI is failing; could you run `tools/test` locally and fix the problems it reports? Some may be fixable automatically with `tools/test --fix`; some will need manual changes.
- Do you know why the stuff in `ios/ZulipMobile.xcodeproj/project.xcworkspace/` was added? I don't understand why it's necessary; perhaps it isn't.
- Please do a pass through your commits to make them [clear and coherent according to our style](https://github.com/zulip/zulip-mobile/blob/main/docs/style.md#git-commits-commit-messages).

**Comment by dizzyjaguar** [comment]
https://media.giphy.com/media/yZBUtAwH4zd9WmTYZ1/giphy.gif

File: .gitignore [hunk 1]
```diff
@@ -3,6 +3,8 @@
 .DS_Store
 
 # Xcode
+*.xcodeproj/
+DerivedData/
```
Comment (chrisbobbe) [inline .gitignore:7]: What happens if you don't add these gitignore lines, and just delete the files they're referring to?

File: src/streams/StreamItem.js [hunk 2]
```diff
@@ -151,13 +161,24 @@ export default function StreamItem(props: Props): Node {
...
     >
       <View style={wrapperStyle}>
+        {isCollapsible === true && (
+          <Touchable onPress={handleCollapse && (() => handleCollapse(streamId))}>
+            <CollapseIcon
+              style={componentStyles.collapseIcon}
+              size={20}
+              color={iconColor}
+              isOpen={isUnCollapsed}
+            />
```
Comment (chrisbobbe) [inline src/streams/StreamItem.js:171]:
```suggestion
            {isUnCollapsed === true ? (
              <IconUnCollapse color={color} size={20} style={componentStyles.collapseIcon} />
            ) : (
              <IconCollapse color={color} size={20} style={componentStyles.collapseIcon} />
            )}
```

I think we don't need this new `CollapseIcon` component to render the icon. The icon and the `Touchable` work closely together to give the user a collapse/uncollapse button. With the icon defined in a separate file, we're asking the reader to jump through the `CollapseIcon` reference just to see how this button is implemented.

**Review by chrisbobbe — CHANGES_REQUESTED** [review]
Thanks, @dizzyjaguar! I've left a few comments about the implementation, below. There are a few smaller things I plan to address in a later review.

In this revision, some of the commits introduce regressions and test failures, which then get fixed in later commits. For example, the first commit adds the collapse icon for _all_ `StreamItem` instances, not just the ones rendered by `UnreadCards`. That gets fixed later, but it would confuse users if we happened cut a release from that first commit. Also, in review, it distracts a bit from the changes you did intend to make. Similarly, `tools/test lint` reports one error and two warnings at that commit. (That particular test output doesn't point to something that users will notice, but test failures often do.)

When this happens, our practice is to not introduce the regressions or test failures in the first place, as part of ["clear and coherent commits"](https://zulip.readthedocs.io/en/latest/contributing/version-control.html). The suggestion I'm about to make should make that easy, as a side effect. 🙂

In this case, I actually think it would make sense to squash all the commits into one. The coherent idea might be summarized as, "`StreamItem` grows a new feature, which one of its callers (the one in `UnreadCards`) starts using." Then when you have that single commit, please give it a properly formatted [commit message](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages).

File: src/unread/UnreadCards.js [hunk 3]
```diff
@@ -59,6 +60,29 @@ export default function UnreadCards(props: Props): Node {
     ...unreadStreamsAndTopics,
   ];
 
+  // move the collpased state of the stream here in a set
+  const [uncollapsedStreamIds, setUncollapsedStreamIds] = useState(new Set<number>());
+
+  // because we are using a Set, which is a mutable, we need to create a newSet to setState to as uncollapsedStreamIds state
```
[the remark this answers is not in the document]
Comment (dizzyjaguar) [reply src/unread/UnreadCards.js:66]: Ah nice this is great! I love this fix, its good to learn about the Immutable library.

File: .gitignore (hunk 1 above)
Comment (dizzyjaguar) [reply .gitignore:7]: Appears to not add the files, I wonder why they were added before in the first place.

**Comment by gnprice** [comment]
> https://media.giphy.com/media/yZBUtAwH4zd9WmTYZ1/giphy.gif

Thanks! Copying that screencap here for a bit more visibility:

giphy

**Comment by gnprice** [comment]
For cross-reference: the issue this fixes is #5423. (I'll edit the description, too, to link the PR to the issue.)

File: src/common/Icons.js [hunk 4]
```diff
@@ -86,6 +86,9 @@ export const IconDiagnostics: SpecificIconType = makeIcon(Feather, 'activity');
 export const IconNotifications: SpecificIconType = makeIcon(Feather, 'bell');
 export const IconLanguage: SpecificIconType = makeIcon(Feather, 'globe');
 export const IconSettings: SpecificIconType = makeIcon(Feather, 'settings');
+// collapsible stream icons
+export const IconUnCollapse: SpecificIconType = makeIcon(FontAwesome, 'caret-up');
+export const IconCollapse: SpecificIconType = makeIcon(FontAwesome, 'caret-down');
```
Comment (gnprice) [inline src/common/Icons.js:91]: These names are potentially pretty confusing because it's hard to be sure if "icon collapse" means
- the icon *to collapse*, shown when the thing is uncollapsed
- the icon *when collapsed*, which you use to uncollapse the thing

Instead `IconCaretUp` and `IconCaretDown` would be good.

Comment (gnprice) [reply src/common/Icons.js:91]: Separately: though these are called "caret", the actual icons seen in your screencap aren't really carets but triangles.

We can get actual carets by using `MaterialIcon` here instead of `Feather`. For a model design, see "Expand and collapse" here:
https://material.io/components/lists

File: src/streams/StreamItem.js [hunk 5]
```diff
@@ -151,13 +160,23 @@ export default function StreamItem(props: Props): Node {
       }}
     >
       <View style={wrapperStyle}>
+        {isCollapsible === true && (
+          <Touchable onPress={handleCollapse && (() => handleCollapse(streamId))}>
```
Comment (gnprice) [inline src/streams/StreamItem.js:164]: One thing `Touchable` does is that when you touch it, it gives visual feedback filling the whole rectangle. (You can see this in the screencap.)

That's appropriate for something styled like a button, but it looks odd here. Instead, we can use `Pressable`. See `LightboxHeader` for a simple example, which makes the "X" icon at the top right of the lightbox screen.

File: src/streams/StreamItem.js [hunk 6]
```diff
@@ -151,13 +160,23 @@ export default function StreamItem(props: Props): Node {
...
+            {isUnCollapsed === true ? (
+              <IconUnCollapse style={componentStyles.collapseIcon} size={20} color={color} />
```
Comment (gnprice) [inline src/streams/StreamItem.js:166]: OTOH a handy thing about that visual feedback -- handy in development, that is -- is that it makes it easy to see from a screencap how far the touchable area extends!

In general, we aim for touch targets of at least 48x48. See discussion here:
https://material.io/design/usability/accessibility.html#layout-and-typography

(For background on the various units for sizes, see here: https://github.com/zulip/zulip-mobile/blob/main/docs/background/layout.md
In short, 1 dp = 1 px = "1" as seen in RN layout props.)

These stream-item rows are less than 48px tall, so the touchable area should extend the full height of the row. When it's smaller, it's often difficult for people to hit.

Ideally it would also be 48px wide, even though it's less than 48px tall.

File: src/streams/StreamItem.js (hunk 5 above)
Comment (gnprice) [inline src/streams/StreamItem.js:164]: This logic is a little bit complicated. One thing that would simplify it would be to pass "this should show the expand/collapse UI" and "here's the callback for the expand/collapse UI" as a single prop, rather than two props.

That helps because all the information carried by `isCollapsible` is already conveyed in `handleCollapse`: `isCollapsible` should be truthy just when `handleCollapse` is. IOW we only need `handleCollapse`, not a separate `isCollapsible`.

File: src/unread/UnreadCards.js [hunk 7]
```diff
@@ -59,6 +61,22 @@ export default function UnreadCards(props: Props): Node {
     ...unreadStreamsAndTopics,
   ];
 
+  // move the collpased state of the stream here in a Immutable Set
+  const [uncollapsedStreamIds, setUncollapsedStreamIds] = useState(new Immutable.Set<number>());
```
Comment (gnprice) [inline src/unread/UnreadCards.js:65]: This comment doesn't really express any information that isn't in the line of code that follows it. In that situation, best to leave the comment out and let the code speak for itself.

**Review by gnprice — COMMENTED** [review]
Thanks @dizzyjaguar, and thanks @chrisbobbe for the previous review!

This is looking like a good direction. Specific comments below.

File: src/unread/UnreadCards.js [hunk 8]
```diff
@@ -59,6 +61,22 @@ export default function UnreadCards(props: Props): Node {
...
+  const handleCollapse = (id: number) => {
+    setUncollapsedStreamIds(
+      uncollapsedStreamIds.has(id) ? uncollapsedStreamIds.delete(id) : uncollapsedStreamIds.add(id),
+    );
+  };
+
+  // if you want all the streams to be open as the default state then add the unreadcards to the uncollapsedStreamIds Set when the component mounts via
+  // useEffect(() => {
+  //   if(unreadCards) {
+  //     unreadCards.forEach(card => uncollapsedStreamIds.add(card.streamId))
```
[the remark this answers is not in the document]
Comment (dizzyjaguar) [reply src/unread/UnreadCards.js:76]: I ended up changing the state to keep track of collapsedStreams rather than expandedStreams, so that should take care of this right?

File: src/common/Icons.js (hunk 4 above)
Comment (dizzyjaguar) [reply src/common/Icons.js:91]: Here is a screen shot with those icons from Material:

Screen Shot 2022-09-02 at 10 40 42 AM

File: src/unread/UnreadCards.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (gnprice) [reply src/unread/UnreadCards.js:76]: Yeah -- depending on how the rest of the code works in the new version, that will probably take care of this.

File: src/streams/StreamItem.js (hunk 6 above)
Comment (dizzyjaguar) [reply src/streams/StreamItem.js:166]: - Here is what the `StreamItem` looks like with the width of the `Touchable` at 48.

Screen Shot 2022-09-04 at 9 54 02 AM

- Currently the surrounding `View` has padding added to it from `listItem` in `miscStyles.js`. With that padding there, if the user clicks in the padding area that is just outside the `Pressable` it will take them to the individual Stream instead of collapse/expand. Should I approach a way to conditionally remove this padding in the area where the `Pressable` is if the StreamItem is one that will be collapsible? 
- If so this should also bring the `Caret` 16px to the left which may make it look a little less wide.

**Comment by gnprice** [comment]
Replying here for visibility to [a subthread above](https://github.com/zulip/zulip-mobile/pull/5467#discussion_r962343879):

> * Here is what the `StreamItem` looks like with the width of the `Touchable` at 48.
> 
> Screen Shot 2022-09-04 at 9 54 02 AM
> 
> * Currently the surrounding `View` has padding added to it from `listItem` in `miscStyles.js`. With that padding there, if the user clicks in the padding area that is just outside the `Pressable` it will take them to the individual Stream instead of collapse/expand. Should I approach a way to conditionally remove this padding in the area where the `Pressable` is if the StreamItem is one that will be collapsible?
> * If so this should also bring the `Caret` 16px to the left which may make it look a little less wide.

Yeah, I agree -- that screenshot looks to have too much horizontal padding, and it'd be good to remove the excess.

Looking forward to seeing the next revision! If you've addressed the other feedback as best you can and there's just an item or two like this one where you have questions, it's helpful to go ahead and send the new revision, just being explicit about the points you're unsure about. More here:
https://zulip.readthedocs.io/en/latest/contributing/contributing.html#how-to-help-move-the-review-process-forward

File: src/streams/StreamItem.js (hunk 6 above)
Comment (gnprice) [reply src/streams/StreamItem.js:166]: (Replied [in the main thread](https://github.com/zulip/zulip-mobile/pull/5467#issuecomment-1238712214) below.)

**Comment by dizzyjaguar** [comment]
@gnprice  I just pushed up the recent revision that has fixed all the suggested changes besides the new issue. https://github.com/zulip/zulip-mobile/pull/5467#discussion_r962343879
(dealing with the excess of padding when the `caret` is present in the `StreamItem`)

File: src/unread/UnreadCards.js [hunk 9]
```diff
@@ -59,6 +61,14 @@ export default function UnreadCards(props: Props): Node {
     ...unreadStreamsAndTopics,
   ];
 
+  const [collapsedStreamIds, setCollapsedStreamIds] = useState(new Immutable.Set<number>());
```
Comment (gnprice) [inline src/unread/UnreadCards.js:64]: `Immutable` containers are initialized with a plain function call, without `new`. See `git grep Immutable` for examples in our codebase.

File: src/streams/StreamItem.js [hunk 10]
```diff
@@ -151,13 +164,26 @@ export default function StreamItem(props: Props): Node {
       }}
     >
       <View style={wrapperStyle}>
+        {handleExpandCollapse && (
+          <Pressable
+            style={componentStyles.touchable}
+            onPress={handleExpandCollapse && (() => handleExpandCollapse(streamId))}
```
Comment (gnprice) [inline src/streams/StreamItem.js:170]: This `handleExpandCollapse &&` is happily now redundant with the one above. That simplification is one of the benefits of unifying that prop with the `isCollapsible` of the previous revision.

File: src/streams/StreamItem.js [hunk 11]
```diff
@@ -47,19 +47,30 @@ const componentStyles = createStyleSheet({
   muted: {
     opacity: 0.5,
   },
+  collapseIcon: {
+    marginRight: 8,
+  },
+  touchable: {
+    // backgroundColor: 'green',
```
Comment (gnprice) [inline src/streams/StreamItem.js:54]: Generally we avoid checking in commented-out code.

(I'm guessing this was here for debugging.)

File: src/streams/StreamItem.js [hunk 12]
```diff
@@ -47,19 +47,30 @@ const componentStyles = createStyleSheet({
   muted: {
     opacity: 0.5,
   },
+  collapseIcon: {
+    marginRight: 8,
+  },
+  touchable: {
```
Comment (gnprice) [inline src/streams/StreamItem.js:53]: nit: `pressable` would be a better name, given this is now a `Pressable` component and not a `Touchable`.

File: src/unread/UnreadCards.js [hunk 13]
```diff
@@ -51,6 +52,7 @@ export default function UnreadCards(props: Props): Node {
   type Card =
     | UnreadStreamItem
     | {| key: 'private', data: $ReadOnlyArray<React$ElementConfig<typeof PmConversationList>> |};
+
```
Comment (gnprice) [inline src/unread/UnreadCards.js:55]: nit: this change isn't related to what the commit is about

**Review by gnprice — COMMENTED** [review]
Thanks @dizzyjaguar! I've looked through the code in this revision, and it generally looks good -- just small comments below.

Looking forward also to seeing a revision addressing that layout issue.

File: src/unread/UnreadCards.js (hunk 13 above)
Comment (dizzyjaguar) [reply src/unread/UnreadCards.js:55]: What would be the best thing to do about this, should I remove this from the commit using a `git rebase -i` ?

File: src/streams/StreamItem.js (hunk 11 above)
Comment (dizzyjaguar) [reply src/streams/StreamItem.js:54]: oops!

File: src/streams/StreamItem.js (hunk 10 above)
Comment (dizzyjaguar) [reply src/streams/StreamItem.js:170]: @gnprice But don't we still need this conditional rendering of the `Pressable caret` since `StreamItem` gets called a few other places in the code. If thats taken away then wouldn't the `caret` always get rendered with `StreamItem`, which may not be preferred?

Here is what the second tab(screens) looks like if I remove that conditional. 

Screen Shot 2022-09-08 at 8 51 13 AM

File: src/unread/UnreadCards.js (hunk 13 above)
Comment (gnprice) [reply src/unread/UnreadCards.js:55]: Yeah, just remove the change from the commit.

File: src/streams/StreamItem.js (hunk 10 above)
Comment (gnprice) [reply src/streams/StreamItem.js:173]: Right, I mean that because the first conditional (around the whole `Pressable`) is there, the conditional on this line (in the value of `onPress`) is redundant.

Comment (dizzyjaguar) [reply src/streams/StreamItem.js:173]: Ahh oops I missed that! That makes more sense haha. Thanks for the clarification!

**Comment by dizzyjaguar** [comment]
@gnprice  Just pushed up this recent revision that cleaned up those small issues as-well as included fixing https://github.com/zulip/zulip-mobile/pull/5467#discussion_r962343879

**Comment by gnprice** [comment]
Thanks @dizzyjaguar for the revision! I think at this point everything looks good, except for the way this handles layout.
* First, a process point: when posting a revision that changes how the UI looks, it's important to post screenshots so people can see the changes and give feedback. Just like it is for posting a PR in the first place: https://github.com/zulip/zulip-mobile/pull/5467#issuecomment-1218428590

Here's a screenshot: 
image

* Looking at that layout, it doesn't look properly aligned to me: there should be less space between the caret and the stream icon "#".

* The implementation of the layout looks brittle, in a couple of respects:
  ```js
    marginLeft: -16,
    marginTop: -12,
    marginBottom: -12,
    width: 48,
    height: 33,
  ```

  * The margins are negative. That's a counterintuitive way for things to work, and makes the layout hard to reason about. Instead of using negative margins to try to cancel out padding elsewhere that we turn out not to want, it'd be better to avoid having padding that we don't want.

    That could involve making the padding conditional. Or it could involve moving the padding so that instead of applying to the enclosing container, it applies to each of the individual items in the container -- then this one can simply not include the padding in the first place.

  * The height is hard-coded to 33px. It's not clear where that number comes from -- I'm guessing it's the result of trying several numbers and tuning it until it looked about right.

    The main problem with that approach is that because it doesn't express our actual intent, it's prone to breaking when things change. That change could be some future change we make elsewhere in the code… or it could be simply someone running the app on a device with different settings, for example with their font size turned up to large for readability.

    On a test device, using the "Element Inspector" from the RN debug menu, I find the height of a StreamItem is 36.7px with default settings. With font size made small, it's down to 34.2px -- though even then, that's taller than 33px. With font size made large, it's 42.5px, much taller.

    (Which illustrates the secondary problem with hardcoding a number like this -- it's likely to be not quite accurate even in the case you were looking at.)

    And in fact, with large font size, if I try to expand or collapse a row and I touch a bit above or below the caret, it's quite easy to end up activating the row as a whole and opening up the stream's message list instead, which could be pretty frustrating.

**Comment by gnprice** [comment]
Oh, and for getting the size of the pressable area right, here's a debugging tip -- set [`android_ripple` on `Pressable`](https://reactnative.dev/docs/0.67/pressable#android_ripple-android):
```diff
           <Pressable
             style={componentStyles.pressable}
             onPress={() => handleExpandCollapse(streamId)}
+            android_ripple={{ color: 'black' }}
           >
```

That gives it visual feedback on Android, which as discussed above:
https://github.com/zulip/zulip-mobile/pull/5467#discussion_r961250603
https://github.com/zulip/zulip-mobile/pull/5467#discussion_r961251756
is undesirable for this bit of UI but is handy in development for seeing exactly where the pressable area extends.

**Comment by alya** [comment]
@dizzyjaguar Just to check in here, do you plan to continue work on this PR? Thanks!

**Comment by dizzyjaguar** [comment]
@alya  Hey yes I do plan to continue, I was just catching up on some things. I'm hoping to start on these corrections this week!

**Comment by gnprice** [comment]
Thanks @dizzyjaguar for the revision!

That layout, and the code, look fine.

However, I just tried this out and found it has a very serious performance problem:
* When I have a stream with a lot of unread topics, and hit the button to collapse the stream, the area below it can remain blank for a long time -- like 10 seconds -- before the next stream appears and starts to fill in with topics.
* Then those topics appear slowly, one by one, at a rate of a handful of them per second.
* If the next stream is already collapsed too, then after it appears the area under it can also remain blank for several more seconds before the next stream appears.

I think the short of what's happening is that `SectionList` (and its underlying `VirtualizedList`) is not prepared for lots and lots of list items suddenly turning to zero-height `null` elements, and doesn't respond to that case efficiently. Unfortunately that means that fixing things so that this has usable performance may not be an approachable issue.

I'll spend a bit more time right now digging into what's happening in the `VirtualizedList` to see if I can figure out more.

**Comment by gnprice** [comment]
OK, here's what I've learned.

First, `VirtualizedList` is making a messy and rather underspecified set of assumptions about the items in the list. Roughly, it assumes that they're all around the same height.

To learn about what it's doing, I added a variety of logging statements. Here's the current full set of them -- it makes pretty verbose results, and I did most of the debugging before adding some of the more verbose of these, but just dumping them all to make it relatively easy to pick up where I left off:
<details>

```diff
diff --git a/../mobile2/node_modules/react-native/Libraries/Lists/VirtualizeUtils.js b/../mobile/node_modules/react-native/Libraries/Lists/VirtualizeUtils.js
index 1d80b378a..f075594b9 100644
--- a/../mobile2/node_modules/react-native/Libraries/Lists/VirtualizeUtils.js
+++ b/../mobile/node_modules/react-native/Libraries/Lists/VirtualizeUtils.js
@@ -117,6 +117,7 @@ export function computeWindowedRenderLimits(
 } {
   const itemCount = getItemCount(data);
   if (itemCount === 0) {
+    console.log('computeWindowedRenderLimits trivial');
     return prev;
   }
   const {offset, velocity, visibleLength} = scrollMetrics;
@@ -142,6 +143,7 @@ export function computeWindowedRenderLimits(
 
   const lastItemOffset = getFrameMetricsApprox(itemCount - 1).offset;
   if (lastItemOffset < overscanBegin) {
+    console.log('computeWindowedRenderLimits at end');
     // Entire list is before our overscan window
     return {
       first: Math.max(0, itemCount - 1 - maxToRenderPerBatch),
@@ -170,25 +172,33 @@ export function computeWindowedRenderLimits(
   // render.
   let newCellCount = newRangeCount(prev, visible);
 
+  console.log('computeWindowedRenderLimits',
+    itemCount, lastItemOffset,
+    {visibleBegin, visibleEnd, overscanBegin, overscanEnd},
+    {first, last, overscanFirst, overscanLast},
+  );
   while (true) {
     if (first <= overscanFirst && last >= overscanLast) {
+      console.log('computeWindowedRenderLimits→ full', first, last, newCellCount);
       // If we fill the entire overscan range, we're done.
       break;
     }
     const maxNewCells = newCellCount >= maxToRenderPerBatch;
     const firstWillAddMore = first <= prev.first || first > prev.last;
     const firstShouldIncrement =
       first > overscanFirst && (!maxNewCells || !firstWillAddMore);
     const lastWillAddMore = last >= prev.last || last < prev.first;
     const lastShouldIncrement =
       last < overscanLast && (!maxNewCells || !lastWillAddMore);
     if (maxNewCells && !firstShouldIncrement && !lastShouldIncrement) {
+      console.log('computeWindowedRenderLimits→ batch', first, last, newCellCount);
       // We only want to stop if we've hit maxNewCells AND we cannot increment first or last
       // without rendering new items. This let's us preserve as many already rendered items as
       // possible, reducing render churn and keeping the rendered overscan range as large as
       // possible.
       break;
     }
+    console.log('computeWindowedRenderLimits…', first, last, newCellCount);
     if (
       firstShouldIncrement &&
       !(fillPreference === 'after' && lastShouldIncrement && lastWillAddMore)
diff --git a/../mobile2/node_modules/react-native/Libraries/Lists/VirtualizedList.js b/../mobile/node_modules/react-native/Libraries/Lists/VirtualizedList.js
index 2648cc3e4..662125fdd 100644
--- a/../mobile2/node_modules/react-native/Libraries/Lists/VirtualizedList.js
+++ b/../mobile/node_modules/react-native/Libraries/Lists/VirtualizedList.js
@@ -1665,6 +1665,10 @@ class VirtualizedList extends React.PureComponent<Props, State> {
         distBottom < 0 ||
         (velocity > 2 && distBottom < scrollingThreshold);
     }
+    console.log('_scheduleCellsToRenderUpdate',
+      this.state, this._scrollMetrics,
+      {itemCount, onEndReachedThreshold, scrollingThreshold, hiPri}
+    );
     // Only trigger high-priority updates if we've actually rendered cells,
     // and with that size estimate, accurately compute how many cells we should render.
     // Otherwise, it would just render as many cells as it can (of zero dimension),
@@ -1733,6 +1737,7 @@ class VirtualizedList extends React.PureComponent<Props, State> {
       getItemCount,
       onEndReachedThreshold: _onEndReachedThreshold,
     } = this.props;
+    console.log('_updateCellsToRender');
     const onEndReachedThreshold = onEndReachedThresholdOrDefault(
       _onEndReachedThreshold,
     );
@@ -1755,15 +1760,17 @@ class VirtualizedList extends React.PureComponent<Props, State> {
           // So let's wait until we've scrolled the view to the right place. And until then,
           // we will trust the initialScrollIndex suggestion.
           if (!this.props.initialScrollIndex || this._scrollMetrics.offset) {
+            const ts = Date.now();
             newState = computeWindowedRenderLimits(
               this.props.data,
               this.props.getItemCount,
               maxToRenderPerBatchOrDefault(this.props.maxToRenderPerBatch),
               windowSizeOrDefault(this.props.windowSize),
               state,
               this._getFrameMetricsApprox,
               this._scrollMetrics,
             );
+            console.log('time computeWindowedRenderLimits', Date.now() - ts);
           }
         }
       } else {
@@ -1837,6 +1844,7 @@ class VirtualizedList extends React.PureComponent<Props, State> {
       // check for invalid frames due to row re-ordering
       return frame;
     } else {
+      console.log('approx frame metrics at', index, 'avg:', this._averageCellLength);
       const {getItemLayout} = this.props;
       invariant(
         !getItemLayout,
```

</details>

I think it *might* suffice to supply a `getItemLayout` prop.
* This would tell `VirtualizedList` exactly how tall each item is, and also how far it is from the top i.e. the sum of how tall everything above it is.
* We'd use the fact that we know all the topics are the same height, except when they're zero height, and that we know which are zero height, to compute this efficiently.
* Writing that is a bit dodgy because it requires knowing the *actual* exact height of the topic items, not just that they're all the same. (Also of all the other items in the list: the `StreamItem` headings, and `PmConversationList` which we shoehorn in as the first item in the list.)
* It's also a bit dodgy because it's not really a clean part of the interface of `SectionList`. It belongs to the interface of `VirtualizedList`, but `SectionList` (really `VirtualizedSectionList` underneath it) adds some complications in how it talks to `VirtualizedList` underneath. So the implementation of `getItemLayout` would need to take care to align with the conventions of the `_getItem` method on `VirtualizedSectionList`.
  * This is probably OK in practice because the last time those conventions changed was in 2017, in facebook/react-native@f702cbecba6248394f9bb660aa834c1b27f3bedd released in RN v0.46.

If we did that, it would completely remove the one place where `VirtualizedList` explicitly assumes that the items are all roughly the same height, in its `_getFrameMetricsApprox` where it takes an average height and uses that to estimate things. OTOH there are other places where it implicitly makes that assumption in the design of its algorithms, which would remain.

But I suspect that will not in fact suffice to solve the problem. In some of my testing, it decides it needs to render a large swath of items… and then it seems to promptly interrupt itself and try for a just slightly larger swath, and then a slightly larger, again and again for many seconds:
```
 LOG  _scheduleCellsToRenderUpdate {"first": 0, "last": 0} {"contentLength": 0, "dOffset": 0, "dt": 10, "offset": 0, "timestamp": 0, "velocity": 0, "visibleLength": 604.727294921875} {"hiPri": false, "itemCount": 0, "onEndReachedThreshold": 2, "scrollingThreshold": 604.727294921875}
 LOG  _scheduleCellsToRenderUpdate {"first": 0, "last": 3419} {"contentLength": 36.3636360168457, "dOffset": 0, "dt": 10, "offset": 0, "timestamp": 0, "velocity": 0, "visibleLength": 628} {"hiPri": true, "itemCount": 3843, "onEndReachedThreshold": 2, "scrollingThreshold": 628}
 LOG  _scheduleCellsToRenderUpdate {"first": 0, "last": 3419} {"contentLength": 36.3636360168457, "dOffset": 0, "dt": 10, "offset": 0, "timestamp": 0, "velocity": 0, "visibleLength": 628} {"hiPri": true, "itemCount": 3843, "onEndReachedThreshold": 2, "scrollingThreshold": 628}
 LOG  _updateCellsToRender
 LOG  computeWindowedRenderLimits 3843 702.0557265161868 {"overscanBegin": 0, "overscanEnd": 6908, "visibleBegin": 0, "visibleEnd": 628} {"first": 0, "last": 3436, "overscanFirst": 0, "overscanLast": 3842}
 LOG  computeWindowedRenderLimits→ batch 0 3436 17
 LOG  time computeWindowedRenderLimits 32
 LOG  _scheduleCellsToRenderUpdate {"first": 0, "last": 0} {"contentLength": 0, "dOffset": 0, "dt": 10, "offset": 0, "timestamp": 0, "velocity": 0, "visibleLength": 604.727294921875} {"hiPri": false, "itemCount": 0, "onEndReachedThreshold": 2, "scrollingThreshold": 604.727294921875}
 LOG  _scheduleCellsToRenderUpdate {"first": 0, "last": 3436} {"contentLength": 36.3636360168457, "dOffset": 0, "dt": 10, "offset": 0, "timestamp": 0, "velocity": 0, "visibleLength": 628} {"hiPri": true, "itemCount": 3843, "onEndReachedThreshold": 2, "scrollingThreshold": 628}
 LOG  _scheduleCellsToRenderUpdate {"first": 0, "last": 3436} {"contentLength": 36.3636360168457, "dOffset": 0, "dt": 10, "offset": 0, "timestamp": 0, "velocity": 0, "visibleLength": 628} {"hiPri": true, "itemCount": 3843, "onEndReachedThreshold": 2, "scrollingThreshold": 628}
 LOG  _updateCellsToRender
 LOG  computeWindowedRenderLimits 3843 698.5454478836059 {"overscanBegin": 0, "overscanEnd": 6908, "visibleBegin": 0, "visibleEnd": 628} {"first": 0, "last": 3454, "overscanFirst": 0, "overscanLast": 3842}
 LOG  computeWindowedRenderLimits→ batch 0 3454 18
 LOG  time computeWindowedRenderLimits 31
 LOG  _scheduleCellsToRenderUpdate {"first": 0, "last": 0} {"contentLength": 0, "dOffset": 0, "dt": 10, "offset": 0, "timestamp": 0, "velocity": 0, "visibleLength": 604.727294921875} {"hiPri": false, "itemCount": 0, "onEndReachedThreshold": 2, "scrollingThreshold": 604.727294921875}
 LOG  _scheduleCellsToRenderUpdate {"first": 0, "last": 3454} {"contentLength": 36.3636360168457, "dOffset": 0, "dt": 10, "offset": 0, "timestamp": 0, "velocity": 0, "visibleLength": 628} {"hiPri": true, "itemCount": 3843, "onEndReachedThreshold": 2, "scrollingThreshold": 628}
 LOG  _scheduleCellsToRenderUpdate {"first": 0, "last": 3454} {"contentLength": 36.3636360168457, "dOffset": 0, "dt": 10, "offset": 0, "timestamp": 0, "velocity": 0, "visibleLength": 628} {"hiPri": true, "itemCount": 3843, "onEndReachedThreshold": 2, "scrollingThreshold": 628}
 LOG  _updateCellsToRender
 LOG  computeWindowedRenderLimits 3843 695.0700973966228 {"overscanBegin": 0, "overscanEnd": 6908, "visibleBegin": 0, "visibleEnd": 628} {"first": 0, "last": 3471, "overscanFirst": 0, "overscanLast": 3842}
 LOG  computeWindowedRenderLimits→ batch 0 3471 17
 LOG  time computeWindowedRenderLimits 32
[… etc etc etc …]
```
It'd be much better off if it just stuck to any one of those and finished it. I'm not sure why it does this, and it's possible that a `getItemLayout` would cause this trap not to spring, but it's not clear that it would.

**Comment by gnprice** [comment]
OK, or here's a fix which works pretty well:
```diff
--- a/src/unread/UnreadCards.js
+++ b/src/unread/UnreadCards.js
@@ -49,19 +49,22 @@ export default function UnreadCards(props: Props): Node {
   const dispatch = useDispatch();
   const conversations = useSelector(getUnreadConversations);
   const unreadStreamsAndTopics = useSelector(getUnreadStreamsAndTopicsSansMuted);
+
+  const [collapsedStreamIds, setCollapsedStreamIds] = useState(Immutable.Set<number>());
+
   type Card =
     | UnreadStreamItem
     | {| key: 'private', data: $ReadOnlyArray<React$ElementConfig<typeof PmConversationList>> |};
   const unreadCards: $ReadOnlyArray<Card> = [
     {
       key: 'private',
       data: [{ conversations }],
     },
-    ...unreadStreamsAndTopics,
+    ...unreadStreamsAndTopics.map(section =>
+      collapsedStreamIds.has(section.streamId) ? { ...section, data: [] } : section,
+    ),
   ];
 
-  const [collapsedStreamIds, setCollapsedStreamIds] = useState(Immutable.Set<number>());
-
   const handleExpandCollapse = (id: number) => {
     setCollapsedStreamIds(
       collapsedStreamIds.has(id) ? collapsedStreamIds.delete(id) : collapsedStreamIds.add(id),
@@ -102,7 +105,7 @@ export default function UnreadCards(props: Props): Node {
       renderItem={({ item, section }) =>
         section.key === 'private' ? (
           <PmConversationList {...item} />
-        ) : collapsedStreamIds.has(section.streamId) ? null : (
+        ) : (
           <TopicItem
             streamId={section.streamId}
             name={item.topic}
```

In other words: when a stream is collapsed, instead of replacing all its topic items with nulls, remove them from the list entirely. That way, we maintain the property that the items are all roughly the same height.

(All except the `PmConversationList` at the top, which can be zero height if there are no unread PMs or much taller than the other items otherwise. But at least we don't have long runs of zero-height items, which is I think what really sends the `VirtualizedList` algorithms over the edge.)

With that fix, the performance still isn't *good* -- it can take a couple of seconds to respond on collapse/uncollapse -- but I think it's acceptable. This whole screen already doesn't perform great when you have tons of unreads and try to scroll quickly, which is the alternative to collapse/uncollapse from a user perspective.

It's getting late here, so I won't try to clean that up for merge now. I'll aim to do so in the next couple of days.

**Comment by gnprice** [comment]
OK, cleaned that fix up and made it an additional commit. (There's a bit of a tension where we might want to squash it instead for the sake of [coherent commits](https://github.com/zulip/zulip-mobile/blob/main/docs/style.md#commit-series); but in this case the fix reflects an interesting substantive issue, so it's helpful to have it separated out explicitly with its own commit message; and also the original commit didn't quite *break* anything, just made things potentially very slow if you actually use the new feature it added, so it's tolerable to have it in the history before the fix.)

Also added a couple of stylistic fixes, as separate commits for visibility, and tweaked the commit message -- do take a look at the whole series:
76df54f5c StreamItem [nfc]: Move styles into render function
0f1fb0e74 unread: Make each stream's section collapsible
729a2892d unread: Fix VirtualizedList performance on un/collapse of stream
2ae97d6d8 icons [nfc]: Cut superfluous comment
a0d0623a6 StreamItem [nfc]: Bring related styles next to each other

Now merging!

**Comment by dizzyjaguar** [comment]
@gnprice @chrisbobbe  Woohoo! 😄 
Thank you both for your pull request reviews and comments. 
It's being really helpful for my own development to work with and learn from such an awesome team of engineers!
Excited to keep contributing to the project 🙌🏼
